### PR TITLE
Implement docker file

### PR DESCRIPTION
**Pull request by WatsonThink**
Implement #177

=Modification= 
Simple docker image of rust environment where the server and the client can run.
Write document of docker.

=Confirmation=
Build image
Run the server and client container
The server and client container can connect

### CODE REVIEW

**Comment by ysimonson** [comment]
A `.dockerignore` would help speed up build times, especially on systems like mine with a lot of local prior build history. These entries seem to be a good minimization of the build context size:

```
/target
/lib/target
/lib/fuzz/artifacts
/lib/fuzz/corpus
/.git
/.github
```

**Comment by ysimonson** [comment]
I think we could make things easier for docker users by building the the client and server ahead of time in the images.

**Comment by WatsonThink** [comment]
Thanks for the advice.
Which for is this docker development or production?
If it is for production, it is good to use multi-stage builds to reduce the image size.

**Comment by ysimonson** [comment]
Production images would have greater bang-for-the-buck. Multi-stage makes sense.

**Comment by WatsonThink** [comment]
I used a multi-stage builds to reduce the image size for the product.
If any changes will be required, let me know.

**Comment by ysimonson** [comment]
At the bottom of the `Dockerfile`, add `CMD` and `EXPOSE` entries so that someone can start the server without running a separate command in shell:

```Dockerfile
CMD ["indradb-server", "-a", "0.0.0.0:27615"]
EXPOSE 27615
```

**Comment by WatsonThink** [comment]
I wrote "CMD ["/bin/bash"]" because this image is shared by a server and a client.
If we make the image with server automatically running, I think we need to make the two images separately and the CMD of client should be /bin/bash.
Does it make sense?

**Comment by ysimonson** [comment]
Yeah that makes sense. It's a little less efficient, but nicer particularly for automation purposes. For the client, you could use `ENTRYPOINT`, which makes it easy to pass arguments in: https://stackoverflow.com/a/40312311

**Comment by WatsonThink** [comment]
I also apply ENTORYPOINT to the server image because it can take arguments.
One problem is a builder image will be redundant,
It can be removed using `docker image prune` but if there is a efficient way, I will fix it.

File: Dockerfile.server [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+RUN apt-get update && apt-get -y install clang && \
+    rm -rf /var/lib/apt/lists/* && \
+    rustup component add rustfmt && \
+    cargo build --lib --bin indradb-server --release
+
+FROM debian:stretch-slim
+WORKDIR /root
+COPY --from=builder /usr/app/src/target/release/indradb-server .
+EXPOSE 27615
+ENTRYPOINT ["./indradb-server"]
```
Comment (ysimonson) [inline Dockerfile.server:13]: I _think_ (but not sure) you could combine these Dockerfiles into one, and create multiple images; e.g. [see here](https://adilsoncarvalho.com/creating-multiple-images-from-a-single-dockerfile-3f69254b6137)

File: Dockerfile.server [hunk 2]
```diff
@@ -0,0 +1,13 @@
+FROM rust:latest as builder
+WORKDIR /usr/app/src
+COPY ./ /usr/app/src
+RUN apt-get update && apt-get -y install clang && \
+    rm -rf /var/lib/apt/lists/* && \
+    rustup component add rustfmt && \
```
Comment (ysimonson) [inline Dockerfile.server:6]: Does `rustfmt` need to be installed?

File: Dockerfile.server [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    cargo build --lib --bin indradb-server --release
```
Comment (ysimonson) [inline Dockerfile.server:7]: Running `cargo build` should be a separate step from the above, to better take advantage of caching. Also, you could use buildkit to speed this up:

```
RUN --mount=type=cache,target=/usr/local/cargo/registry \
  --mount=type=cache,target=/rust/target \
  cargo build --release &&\
  cp /rust/target/release/test /rust/test
```

With that, the entire thing won't be recompiled from scratch on every `docker build`.

File: Dockerfile.server (hunk 2 above)
Comment (WatsonThink) [reply Dockerfile.server]: If `rustfmt` is not in the image, the following error occurs when building

 ```
 => [builder 5/5] RUN --mount=type=cache,target=/usr/local/cargo/registry     --mount=type=cache,target=/rust/t  308.2s
 => => #   process didn't exit successfully: `/usr/app/src/target/release/build/indradb-proto-9d40d23f8a968169/build-sc
 => => # ript-build` (exit code: 1)
 => => #   --- stderr
 => => #   error: 'rustfmt' is not installed for the toolchain '1.51.0-x86_64-unknown-linux-gnu'
 => => #   To install, run `rustup component add rustfmt`
 => => # warning: build failed, waiting for other jobs to finish...
```

**Comment by WatsonThink** [comment]
I integrated the client and the server image and use the buildkit to speed up building.

**Comment by ysimonson** [comment]
Just tested it out, everything looks good! Thanks again for this contribution.

**Pull request merged**